diff --git a/squid.inc b/squid.inc
index bd0c863..20f0eee 100755
--- squid.inc
+++ squid.inc
@@ -498,7 +498,7 @@ function squid_validate_auth($post, $input_errors) {
 	}
 
 	$auth_method = $post['auth_method'];
-	if (($auth_method != 'none') && ($auth_method != 'local')) {
+	if (($auth_method != 'none') && ($auth_method != 'local') && ($auth_method != 'ntlm')) {
 		$server = trim($post['auth_server']);
 		if (empty($server))
 			$input_errors[] = 'The field \'Authentication server\' is required';
@@ -1069,6 +1069,15 @@ function squid_resync_auth() {
 				$conf .= "auth_param basic program " . SQUID_LOCALBASE . "/libexec/squid/msnt_auth\n";
 				squid_resync_msnt();
 				break;
+                        case 'ntlm':
+                                if ($settings['ntlmssp'] == 'on') {
+                                        $domain_samba4 = (($settings['ntlm_domain'] <> "") ? "--domain={$settings['ntlm_domain']} " : "");
+                                        $ntlm_proc = (isset($settings['ntlm_proc']) ? "{$settings['ntlm_proc']}" : "20");
+                                        $conf .= "auth_param ntlm program /usr/pbi/squid-amd64/libexec/squid/ntlm_auth " . $domain_samba4 . "--helper-protocol=squid-2.5-ntlmssp\n";
+                                        $conf .= "auth_param ntlm children {$ntlm_proc}\n";
+                                }
+                                $conf .= "auth_param basic program /usr/pbi/squid-amd64/libexec/squid/ntlm_auth --helper-protocol=squid-2.5-basic\n";
+                                break;
 		}
 		$conf .= <<<EOD
 auth_param basic children $processes
@@ -1212,6 +1221,9 @@ function on_auth_method_changed() {
 	document.iform.auth_ttl.disabled = 1;
 	document.iform.unrestricted_auth.disabled = 1;
 	document.iform.no_auth_hosts.disabled = 1;
+        document.iform.ntlmssp.disabled = 1;
+        document.iform.ntlm_proc.disabled = 1;
+        document.iform.ntlm_domain.disabled = 1;
 }
 -->
 </script>
@@ -1243,6 +1255,9 @@ function on_auth_method_changed() {
 		document.iform.auth_ttl.disabled = 1;
 		document.iform.unrestricted_auth.disabled = 1;
 		document.iform.no_auth_hosts.disabled = 1;
+                document.iform.ntlmssp.disabled = 1;
+                document.iform.ntlm_proc.disabled = 1;
+                document.iform.ntlm_domain.disabled = 1;
 	}
 	else {
 		document.iform.auth_prompt.disabled = 0;
@@ -1265,6 +1280,9 @@ function on_auth_method_changed() {
 			document.iform.ldap_basedomain.disabled = 1;
 			document.iform.radius_secret.disabled = 1;
 			document.iform.msnt_secondary.disabled = 1;
+                        document.iform.ntlmssp.disabled = 1;
+                        document.iform.ntlm_proc.disabled = 1;
+                        document.iform.ntlm_domain.disabled = 1;
 			break;
 		case 'ldap':
 			document.iform.auth_server.disabled = 0;
@@ -1278,6 +1296,9 @@ function on_auth_method_changed() {
 			document.iform.radius_secret.disabled = 1;
 			document.iform.msnt_secondary.disabled = 1;
 			document.iform.auth_ntdomain.disabled = 1;
+                        document.iform.ntlmssp.disabled = 1;
+                        document.iform.ntlm_proc.disabled = 1;
+                        document.iform.ntlm_domain.disabled = 1;
 			break;
 		case 'radius':
 			document.iform.auth_server.disabled = 0;
@@ -1291,6 +1312,9 @@ function on_auth_method_changed() {
 			document.iform.radius_secret.disabled = 0;
 			document.iform.msnt_secondary.disabled = 1;
 			document.iform.auth_ntdomain.disabled = 1;
+                        document.iform.ntlmssp.disabled = 1;
+                        document.iform.ntlm_proc.disabled = 1;
+                        document.iform.ntlm_domain.disabled = 1;
 			break;
 		case 'msnt':
 			document.iform.auth_server.disabled = 0;
@@ -1304,7 +1328,27 @@ function on_auth_method_changed() {
 			document.iform.ldap_basedomain.disabled = 1;
 			document.iform.radius_secret.disabled = 1;
 			document.iform.msnt_secondary.disabled = 0;
+                        document.iform.ntlmssp.disabled = 1;
+                        document.iform.ntlm_proc.disabled = 1;
+                        document.iform.ntlm_domain.disabled = 1;
 			break;
+                case 'ntlm':
+                        document.iform.auth_server.disabled = 1;
+                        document.iform.auth_server_port.disabled = 1;
+                        document.iform.auth_ntdomain.disabled = 1;
+                        document.iform.ldap_user.disabled = 1;
+                        document.iform.ldap_pass.disabled = 1;
+                        document.iform.ldap_version.disabled = 1;
+                        document.iform.ldap_userattribute.disabled = 1;
+                        document.iform.ldap_filter.disabled = 1;
+                        document.iform.ldap_basedomain.disabled = 1;
+                        document.iform.radius_secret.disabled = 1;
+                        document.iform.msnt_secondary.disabled = 1;
+                        document.iform.ntlmssp.disabled = 0;
+                        document.iform.ntlm_proc.disabled = 0;
+                        document.iform.ntlm_domain.disabled = 0;
+                        break;
+
 	}
 }
 -->
diff --git a/squid_auth.xml b/squid_auth.xml
index d28beb0..e2e4536 100755
--- squid_auth.xml
+++ squid_auth.xml
@@ -94,6 +94,7 @@
 				<option><name>LDAP</name><value>ldap</value></option>
 				<option><name>RADIUS</name><value>radius</value></option>
 				<option><name>NT domain</name><value>msnt</value></option>
+                                <option><name>Winbind NTLM</name><value>ntlm</value></option>
 			</options>
 			<onchange>on_auth_method_changed()</onchange>
 		</field>
@@ -180,6 +181,28 @@
 			<type>input</type>
 			<size>60</size>
 		</field>
+                <field>
+                        <fielddescr>NTLMSSP Negotiate</fielddescr>
+                        <fieldname>ntlmssp</fieldname>
+                        <description>Use the NTLMSSP Negotiate packet (Transparent authentication in Windows Domain)</description>
+                        <type>checkbox</type>
+                        <default_value>on</default_value>
+                </field>
+                <field>
+                        <fielddescr>NTLMSSP Auth processes</fielddescr>
+                        <fieldname>ntlm_proc</fieldname>
+                        <description>The number of authenticator processes to spawn. If many authentications are expected within a short timeframe, increase this number accordingly.</description>
+                        <type>input</type>
+                        <size>40</size>
+                        <default_value>20</default_value>
+                </field>
+                <field>
+                        <fielddescr>NTLM Default Domain</fielddescr>
+                        <fieldname>ntlm_domain</fieldname>
+                        <description>Set the default domain for use in ntlm authentication</description>
+                        <type>input</type>
+                        <size>50</size>
+                </field>
 		<field>
 			<fielddescr>Authentication prompt</fielddescr>
 			<fieldname>auth_prompt</fieldname>
