diff --git a/squid_auth.xml b/squid_auth.xml
index e0f2b7b..08b6ae8 100644
--- squid_auth.xml
+++ squid_auth.xml
@@ -105,6 +105,7 @@
 				<option><name>RADIUS</name><value>radius</value></option>
 				<option><name>Captive Portal</name><value>cp</value></option>
 				<option><name>NT Domain</name><value>msnt</value></option>
+				<option><name>Winbind NTLM</name><value>ntlm</value></option>
 			</options>
 			<onchange>on_auth_method_changed()</onchange>
 		</field>
@@ -180,6 +181,32 @@
 			<encoding>base64</encoding>
 		</field>
 		<field>
+		    <name>Squid Authentication NTLM Settings</name>
+		    <type>listtopic</type>
+		</field>
+		<field>
+		    <fielddescr>NTLMSSP Negotiate</fielddescr>
+		    <fieldname>ntlmssp</fieldname>
+		    <description>Use the NTLMSSP Negotiate packet (Transparent authentication in Windows Domain)</description>
+		    <type>checkbox</type>
+		    <default_value>on</default_value>
+		</field>
+		<field>
+		    <fielddescr>NTLMSSP Auth processes</fielddescr>
+		    <fieldname>ntlm_proc</fieldname>
+		    <description>The number of authenticator processes to spawn. If many authentications are expected within a short timeframe, increase this number accordingly.</description>
+		    <type>input</type>
+		    <size>40</size>
+		    <default_value>20</default_value>
+		</field>
+		<field>
+		    <fielddescr>NTLM Default Domain</fielddescr>
+		    <fieldname>ntlm_domain</fieldname>
+		    <description>Set the default domain for use in ntlm authentication</description>
+		    <type>input</type>
+		    <size>50</size>
+		</field>
+		<field>
 			<name>Squid Authentication LDAP Settings</name>
 			<type>listtopic</type>
 		</field>
diff --git a/squid_js.inc b/squid_js.inc
index 33f1923..655428b 100644
--- squid_js.inc
+++ squid_js.inc
@@ -73,6 +73,9 @@ function on_auth_method_changed() {
 	document.iform.auth_ttl.disabled = 1;
 	document.iform.unrestricted_auth.disabled = 1;
 	document.iform.no_auth_hosts.disabled = 1;
+	document.iform.ntlmssp.disabled = 1;
+	document.iform.ntlm_proc.disabled = 1;
+	document.iform.ntlm_domain.disabled = 1;
 }
 //]]>
 </script>
@@ -105,6 +108,9 @@ function on_auth_method_changed() {
 		document.iform.auth_ttl.disabled = 1;
 		document.iform.unrestricted_auth.disabled = 1;
 		document.iform.no_auth_hosts.disabled = 1;
+		document.iform.ntlmssp.disabled = 1;
+		document.iform.ntlm_proc.disabled = 1;
+		ocument.iform.ntlm_domain.disabled = 1;
 	} else {
 		document.iform.auth_prompt.disabled = 0;
 		document.iform.auth_processes.disabled = 0;
@@ -126,6 +132,9 @@ function on_auth_method_changed() {
 		document.iform.ldap_basedomain.disabled = 1;
 		document.iform.radius_secret.disabled = 1;
 		document.iform.msnt_secondary.disabled = 1;
+		document.iform.ntlmssp.disabled = 1;
+		document.iform.ntlm_proc.disabled = 1;
+		document.iform.ntlm_domain.disabled = 1;
 		break;
 	case 'ldap':
 		document.iform.auth_server.disabled = 0;
@@ -139,6 +148,9 @@ function on_auth_method_changed() {
 		document.iform.radius_secret.disabled = 1;
 		document.iform.msnt_secondary.disabled = 1;
 		document.iform.auth_ntdomain.disabled = 1;
+		document.iform.ntlmssp.disabled = 1;
+		document.iform.ntlm_proc.disabled = 1;
+		ocument.iform.ntlm_domain.disabled = 1;
 		break;
 	case 'radius':
 		document.iform.auth_server.disabled = 0;
@@ -152,6 +164,9 @@ function on_auth_method_changed() {
 		document.iform.radius_secret.disabled = 0;
 		document.iform.msnt_secondary.disabled = 1;
 		document.iform.auth_ntdomain.disabled = 1;
+		document.iform.ntlmssp.disabled = 1;
+		document.iform.ntlm_proc.disabled = 1;
+		document.iform.ntlm_domain.disabled = 1;
 		break;
 	case 'msnt':
 		document.iform.auth_server.disabled = 0;
@@ -165,6 +180,9 @@ function on_auth_method_changed() {
 		document.iform.ldap_basedomain.disabled = 1;
 		document.iform.radius_secret.disabled = 1;
 		document.iform.msnt_secondary.disabled = 0;
+		document.iform.ntlmssp.disabled = 1;
+		document.iform.ntlm_proc.disabled = 1;
+		document.iform.ntlm_domain.disabled = 1;
 		break;
 	case 'cp':
 		document.iform.auth_server.disabled = 1;
@@ -183,6 +201,30 @@ function on_auth_method_changed() {
 		document.iform.auth_ttl.disabled = 0;
 		document.iform.unrestricted_auth.disabled = 1;
 		document.iform.no_auth_hosts.disabled = 1;
+		document.iform.ntlmssp.disabled = 1;
+		document.iform.ntlm_proc.disabled = 1;
+		document.iform.ntlm_domain.disabled = 1;
+		break;
+	case 'ntlm':
+		document.iform.auth_server.disabled = 1;
+		document.iform.auth_server_port.disabled = 1;
+		document.iform.auth_ntdomain.disabled = 1;
+		document.iform.ldap_user.disabled = 1;
+		document.iform.ldap_version.disabled = 1;
+		document.iform.ldap_userattribute.disabled = 1;
+		document.iform.ldap_filter.disabled = 1;
+		document.iform.ldap_pass.disabled = 1;
+		document.iform.ldap_basedomain.disabled = 1;
+		document.iform.radius_secret.disabled = 1;
+		document.iform.msnt_secondary.disabled = 1;
+		document.iform.auth_prompt.disabled = 0;
+		document.iform.auth_processes.disabled = 0;
+		document.iform.auth_ttl.disabled = 0;
+		document.iform.unrestricted_auth.disabled = 0;
+		document.iform.no_auth_hosts.disabled = 0;
+		document.iform.ntlmssp.disabled = 0;
+		document.iform.ntlm_proc.disabled = 0;
+		document.iform.ntlm_domain.disabled = 0;
 		break;
 	}
 }