diff --git a/squid.inc b/squid.inc
index 36de09a..9522422 100644
--- squid.inc
+++ squid.inc
@@ -1091,7 +1091,7 @@ function squid_validate_auth($post, &$input_errors) {
 	unset($num_fields);
 
 	$auth_method = $post['auth_method'];
-	if (($auth_method != 'none') && ($auth_method != 'local') && ($auth_method != 'cp')) {
+	if (($auth_method != 'none') && ($auth_method != 'local') && ($auth_method != 'cp') && ($auth_method != 'ntlm')) {
 		$server = trim($post['auth_server']);
 		if (empty($server)) {
 			$input_errors[] = "'Authentication server' is required.";
@@ -1891,6 +1891,16 @@ function squid_resync_auth() {
 				$conf .= "external_acl_type check_cp {$helpers_num} ttl={$auth_ttl} %SRC " . SQUID_BASE . "/bin/check_ip.php\n";
 				$conf .= "acl password external check_cp\n";
 				break;
+			case 'ntlm':
+				if ($settings['ntlmssp'] == 'on') {
+					$domain_samba4 = (($settings['ntlm_domain'] <> "") ? "--domain={$settings['ntlm_domain']} " : "");
+					$ntlm_proc = (isset($settings['ntlm_proc']) ? "{$settings['ntlm_proc']}" : "20");
+					$conf .= "auth_param ntlm program ". SQUID_LOCALBASE . "/libexec/squid/ntlm_auth " . $domain_samba4 . "--helper-protocol=squid-2.5-ntlmssp\n";
+					$conf .= "auth_param ntlm children {$ntlm_proc}\n";
+					$conf .= "auth_param ntlm keep_alive off\n";
+				}
+				$conf .= "auth_param basic program " . SQUID_LOCALBASE . "/libexec/squid/ntlm_auth --helper-protocol=squid-2.5-basic\n";
+				break;
 		}
 		if ($auth_method != 'cp') {
 		$conf .= <<< EOD
diff --git a/squid_auth.xml b/squid_auth.xml
index 2453bde..f84ffb7 100644
--- squid_auth.xml
+++ squid_auth.xml
@@ -104,6 +104,7 @@
 				<option><name>LDAP</name><value>ldap</value></option>
 				<option><name>RADIUS</name><value>radius</value></option>
 				<option><name>Captive Portal</name><value>cp</value></option>
+				<option><name>Winbind NTLM</name><value>ntlm</value></option>
 			</options>
 			<onchange>on_auth_method_changed()</onchange>
 		</field>
@@ -178,6 +179,32 @@
 			<rows>5</rows>
 			<encoding>base64</encoding>
 		</field>
+ 		<field>
+		    <name>Squid Authentication NTLM Settings</name>
+		    <type>listtopic</type>
+		</field>
+		<field>
+		    <fielddescr>NTLMSSP Negotiate</fielddescr>
+		    <fieldname>ntlmssp</fieldname>
+		    <description>Use the NTLMSSP Negotiate packet (Transparent authentication in Windows Domain)</description>
+		    <type>checkbox</type>
+		    <default_value>on</default_value>
+		</field>
+		<field>
+		    <fielddescr>NTLMSSP Auth processes</fielddescr>
+		    <fieldname>ntlm_proc</fieldname>
+		    <description>The number of authenticator processes to spawn. If many authentications are expected within a short timeframe, increase this number accordingly.</description>
+		    <type>input</type>
+		    <size>40</size>
+		    <default_value>20</default_value>
+		</field>
+		<field>
+		    <fielddescr>NTLM Default Domain</fielddescr>
+		    <fieldname>ntlm_domain</fieldname>
+		    <description>Set the default domain for use in ntlm authentication</description>
+		    <type>input</type>
+		    <size>50</size>
+		</field>
 		<field>
 			<name>Squid Authentication LDAP Settings</name>
 			<type>listtopic</type>
diff --git a/squid_js.inc b/squid_js.inc
index 52958ea..235a5ed 100644
--- squid_js.inc
+++ squid_js.inc
@@ -71,6 +71,9 @@ function on_auth_method_changed() {
 	$('#auth_ttl').prop("disabled", true);
 	$('#unrestricted_auth').prop("disabled", true);
 	$('#no_auth_hosts').prop("disabled", true);
+	$('#ntlmssp').prop("disabled", true);
+	$('#ntlm_proc').prop("disabled", true);
+	$('#ntlm_domain').prop("disabled", true);
 }
 //]]>
 </script>
@@ -100,6 +103,9 @@ function on_auth_method_changed() {
 		$('#auth_ttl').prop("disabled", true);
 		$('#unrestricted_auth').prop("disabled", true);
 		$('#no_auth_hosts').prop("disabled", true);
+		$('#ntlmssp').prop("disabled", true);
+		$('#ntlm_proc').prop("disabled", true);
+		$('#ntlm_domain').prop("disabled", true);
 	} else {
 		$('#auth_prompt').prop("disabled", false);
 		$('#auth_processes').prop("disabled", false);
@@ -119,6 +125,9 @@ function on_auth_method_changed() {
 		$('#ldap_filter').prop("disabled", true);
 		$('#ldap_basedomain').prop("disabled", true);
 		$('#radius_secret').prop("disabled", true);
+		$('#ntlmssp').prop("disabled", true);
+		$('#ntlm_proc').prop("disabled", true);
+		$('#ntlm_domain').prop("disabled", true);
 		break;
 	case 'ldap':
 		$('#auth_server').prop("disabled", false);
@@ -130,6 +139,9 @@ function on_auth_method_changed() {
 		$('#ldap_filter').prop("disabled", false);
 		$('#ldap_basedomain').prop("disabled", false);
 		$('#radius_secret').prop("disabled", true);
+		$('#ntlmssp').prop("disabled", true);
+		$('#ntlm_proc').prop("disabled", true);
+		$('#ntlm_domain').prop("disabled", true);
 		break;
 	case 'radius':
 		$('#auth_server').prop("disabled", false);
@@ -141,6 +153,9 @@ function on_auth_method_changed() {
 		$('#ldap_filter').prop("disabled", true);
 		$('#ldap_basedomain').prop("disabled", true);
 		$('#radius_secret').prop("disabled", false);
+		$('#ntlmssp').prop("disabled", true);
+		$('#ntlm_proc').prop("disabled", true);
+		$('#ntlm_domain').prop("disabled", true);
 		break;
 	case 'cp':
 		$('#auth_server').prop("disabled", true);
@@ -157,6 +172,28 @@ function on_auth_method_changed() {
 		$('#auth_ttl').prop("disabled", false);
 		$('#unrestricted_auth').prop("disabled", true);
 		$('#no_auth_hosts').prop("disabled", true);
+		$('#ntlmssp').prop("disabled", true);
+		$('#ntlm_proc').prop("disabled", true);
+		$('#ntlm_domain').prop("disabled", true);
+		break;
+	case 'ntlm':
+		$('#auth_server').prop("disabled", true);
+		$('#auth_server_port').prop("disabled", true);
+		$('#ldap_user').prop("disabled", true);
+		$('#ldap_pass').prop("disabled", true);
+		$('#ldap_version').prop("disabled", true);
+		$('#ldap_userattribute').prop("disabled", true);
+		$('#ldap_filter').prop("disabled", true);
+		$('#ldap_basedomain').prop("disabled", true);
+		$('#radius_secret').prop("disabled", true);
+       $('#auth_prompt').prop("disabled", false);
+       $('#auth_processes').prop("disabled", false);
+       $('#auth_ttl').prop("disabled", false);
+       $('#unrestricted_auth').prop("disabled", false);
+       $('#no_auth_hosts').prop("disabled", false);
+		$('#ntlmssp').prop("disabled", false);
+		$('#ntlm_proc').prop("disabled", false);
+		$('#ntlm_domain').prop("disabled", false);
 		break;
 	}
 }
